### ISSUE TITLE
Prevent throwing error on reset DataGridView current cell address while binding with DataSource

### ISSUE THREAD

**Issue by M-Lipin** [dotnet/winforms#2853]
* .NET Core Version: 
.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-alpha.1.20073.10
 Commit:    29f4d693a9

* Have you experienced this same bug with .NET Framework?: 
Yes

**Problem description:**

The issue is that in some circumstances there is an error when showing the DataGridView control that is already has a DataSource bound. On control show binding is reset which may lead to an error while resetting current cell to -1;-1 coordinates.

An error is shown on form opening / control showing:
_"Operation is not valid because it results in a reentrant call to the SetCurrentCellAddressCore function."_

Stacktrace:
```
System.Windows.Forms!System.Windows.Forms.DataGridView.SetCurrentCellAddressCore in DataGridViewMethods.cs
at System.Windows.Forms!System.Windows.Forms.DataGridView.set_CurrentCell in DataGridView.cs
at System.Windows.Forms!System.Windows.Forms.DataGridView.OnBindingContextChanged in DataGridViewMethods.cs
at System.Windows.Forms!System.Windows.Forms.Control.CreateControl in Control.cs
at System.Windows.Forms!System.Windows.Forms.Control.WmShowWindow in Control.cs
at System.Windows.Forms!System.Windows.Forms.Control.WndProc in Control.cs
at System.Windows.Forms!System.Windows.Forms.DataGridView.WndProc in DataGridViewMethods.cs
```

**Expected behavior:**
There should be no error.

**Minimal repro:**
To be found.

**Comment by M-Lipin** [comment]
Possible solution is to change
`if (dataGridViewOper[DATAGRIDVIEWOPER_inCurrentCellChange] &&                      // Allow the code to be re-entrant only as a result of
    (dataConnection == null || !dataConnection.ProcessingListChangedEvent))   // underlying data changing.
{
    throw new InvalidOperationException(SR.DataGridView_SetCurrentCellAddressCoreNotReentrant);
}`

to

`if (dataGridViewOper[DATAGRIDVIEWOPER_inCurrentCellChange] &&                   // Allow the code to be re-entrant only as a result of
    (dataConnection == null || !dataConnection.ProcessingListChangedEvent) &&  // underlying data changing.​
    !dataGridViewState2[DATAGRIDVIEWSTATE2_inBindingContextChanged])
{​
    throw new InvalidOperationException(SR.DataGridView_SetCurrentCellAddressCoreNotReentrant);​
}`

**Comment by M-Lipin** [comment]
DevDiv issue: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/733640

**Comment by RussKie** [comment]
I think it is sensible, but we need to have a repro

**Comment by merriemcgaw** [comment]
@Shyam-Gupta is investigating the internal Watson ticket

**Comment by Shyam-Gupta** [comment]
The watson bug has been resolved as there were no new hits after 16.2. It seems it has got fixed after DataGridView was replaced with TextBox in String Collection Editor in 16.2. So this bug is not actionable as well.

**Comment by dreddy-work** [comment]
@M-Lipin, referenced watson bug was resolved with no hits in the recent builds.  Unless we have a repro for this scenario, we should close this issue.
